symphony-bdk-core/src/main/java/com/symphony/bdk/core/util/RetryWithRecovery.java

```diff
@@ -0,0 +1,119 @@
+package com.symphony.bdk.core.util;
+
+import com.symphony.bdk.core.api.invoker.ApiException;
+import com.symphony.bdk.core.config.model.BdkRetryConfig;
+
+import io.github.resilience4j.retry.Retry;
+import io.github.resilience4j.retry.RetryConfig;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.Map;
+import java.util.function.Predicate;
+
+/**
+ * This class aims to implement a retry mechanism (on top of a{@link Retry})
+ * with different recovery strategies based on predicates.
+ * @param <T> the type of the object returned by {@link #execute()}
+ */
+@Slf4j
+public class RetryWithRecovery<T> {
+  private SupplierWithApiException<T> supplier;
+  private Predicate<ApiException> ignoreApiException;
+  private Map<Predicate<ApiException>, ConsumerWithThrowable> recoveryStrategies;
+  private Retry retry;
+
+  /**
+   * Constructor with no predicate on when to ignore an {@link ApiException},
+   * i.e. ApiExceptions will never be ignored.
+   * @param name the name of the {@link Retry} service.
+   * @param bdkRetryConfig the retry configuration to be used.
+   * @param supplier the supplier responsible to provide the object of param type T and which may throw an {@link ApiException}.
+   * @param retryOnExceptionPredicate predicate on a thrown {@link ApiException} to know if call should be retried.
+   * @param recoveryStrategies mapping between {@link Predicate<ApiException>} and the corresponding recovery functions to be executed before retrying.
+   *                           If several predicates match, all corresponding consumers will be executed.
+   */
+  public RetryWithRecovery(String name, BdkRetryConfig bdkRetryConfig, SupplierWithApiException<T> supplier,
```

Comment:
I'd prefer to have `RetryFactory` that already hold the `BdkRetryConfig`. This would avoid passing it `RetryWithRecovery` each time we create a new instance